src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs

```diff
@@ -1251,4 +1251,1192 @@ static class C
+                    M(x);
+                }
+            }
+            """;
+        var verifier = CompileAndVerify(source, expectedOutput: "5", verify: Verification.Fails);
+        verifier.VerifyDiagnostics(
+            // (7,11): warning CS9503: Argument 1 should be passed with 'ref' or 'in' keyword
+            //         M(x);
+            Diagnostic(ErrorCode.WRN_ArgExpectedRefOrIn, "x").WithArguments("1").WithLocation(7, 11));
+        verifier.VerifyIL("C.Main", """
+            {
+              // Code size       11 (0xb)
+              .maxstack  1
+              IL_0000:  ldsflda    "int C.x"
+              IL_0005:  call       "void C.M(ref readonly int)"
+              IL_000a:  ret
+            }
+            """);
+    }
+
+    [Fact]
+    public void RefReadonlyParameter_ReadonlyRef_RefArgument()
+    {
+        var source = """
+            class C
+            {
+                static void M(ref readonly int p) => System.Console.Write(p);
+                static readonly int x = 5;
+                static void Main()
+                {
+                    M(ref x);
+                }
+            }
+            """;
+        // PROTOTYPE: Should not be an error when value checks are relaxed. Verify execution and IL.
+        CreateCompilation(source).VerifyDiagnostics(
+            // (7,15): error CS0199: A static readonly field cannot be used as a ref or out value (except in a static constructor)
+            //         M(ref x);
+            Diagnostic(ErrorCode.ERR_RefReadonlyStatic, "x").WithLocation(7, 15));
+    }
+
+    [Fact]
+    public void RefReadonlyParameter_ReadonlyRef_InArgument()
+    {
+        var source = """
+            class C
+            {
+                static void M(ref readonly int p) => System.Console.Write(p);
+                static readonly int x = 5;
+                static void Main()
+                {
+                    M(in x);
+                }
+            }
+            """;
+        var verifier = CompileAndVerify(source, expectedOutput: "5", verify: Verification.Fails);
+        verifier.VerifyDiagnostics();
+        verifier.VerifyIL("C.Main", """
+            {
+              // Code size       11 (0xb)
+              .maxstack  1
+              IL_0000:  ldsflda    "int C.x"
+              IL_0005:  call       "void C.M(ref readonly int)"
+              IL_000a:  ret
+            }
+            """);
+    }
+
+    [Fact]
+    public void RefReadonlyParameter_RValue_PlainArgument()
+    {
+        var source = """
+            class C
+            {
+                static void M(ref readonly int p) => System.Console.Write(p);
+                static void Main()
+                {
+                    M(5);
+                }
+            }
+            """;
+        var verifier = CompileAndVerify(source, expectedOutput: "5");
+        verifier.VerifyDiagnostics(
+            // (6,11): warning CS9503: Argument 1 should be passed with 'ref' or 'in' keyword
```

Comment:
>// (6,11): warning CS9503: Argument 1 should be passed with 'ref' or 'in' keyword [](http://example.com/codeflow?start=12&length=81)

Is the warning going to provide a good guidance for a user in this scenario? Is ```M(ref 5)``` or ```M(in 5)``` going to compile successfully? Effectively, this is the change that the warning is recommending.

Reply:
Yes, this is changed in the follow up PR where value checks are implemented. Adding a prototype comment.

Reply:
> Yes, this is changed in the follow up PR where value checks are implemented. Adding a prototype comment.

I am not sure what is there to change for value checks. Neither `ref` nor `in` should be legal for an rvalue